teps/0096-pipelines-v1-release.md

```diff
@@ -0,0 +1,104 @@
+---
+status: proposed
+title: Pipelines V1 Release
+creation-date: '2021-11-29'
+last-updated: '2021-11-29'
+authors:
+- '@lbernick'
+- '@jerop'
+- '@bobcatfish'
+- '@vdemeester'
+---
+
+# TEP-0096: Pipelines V1 Release
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Background](#background)
+- [Proposal](#proposal)
+  - [V1 Stability and Deprecation Policy](#v1-stability-and-deprecation-policy)
+  - [API Definition](#api-definition)
+  - [Use Cases](#use-cases)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Today, Tekton provides Alpha and Beta APIs.
+This TEP proposes a path to releasing Tekton Pipelines V1 to provide a stable API that users can rely on.
+This TEP documents the API definition and stability policy for Pipelines V1,
+defines the use cases we would like to support for Pipelines V1, and discusses features needed to support these use cases.
+
+## Motivation
+
+To become the industry standard, cloud-native CI/CD platform,
+Tekton must provide users with a stable API they can rely on.
+Some users may be waiting for V1 stability before deciding to use Tekton.
+
+### Goals
+
+- Define stability policy for Pipelines V1.
+- Define supported and unsupported use cases for Pipelines V1.
+- Define criteria for a feature to be "required" or "optional" for Pipelines V1.
+- Define what updates, if any, are needed to the [Pipelines API definition](https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md) before V1 release.
+
+### Non Goals
+
+- Define the stability or V1 plans of other Tekton project.
+- Discuss aspects of Pipelines unrelated to the API, such as performance, dogfooding, release automation,
+and testing and debugging features.
+
+## Background
+
+Tekton currently uses Kubernetes' [API versioning guidelines](https://kubernetes.io/docs/reference/using-api/api-overview/#api-versioning)
+to define API stability levels, as described in the [compatibility policy](https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md#alpha-beta-and-ga).
+
+Tekton API components may contain features at a lower level of stability, gated behind feature flags,
+as described in [TEP-0033: Tekton Feature Gates](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md).
+
+Both Tekton and Kubernetes software releases use [semantic versioning](https://semver.org), but Tekton does not follow 
+Kubernetes' guidelines for [release versioning](https://github.com/kubernetes/sig-release/blob/master/release-engineering/versioning.md).
+Specifically, Tekton does not include release tags containing "alpha", "beta", or "rc".
```

Comment:
I am not sure I understand what is the relation between this (release *tags*) and api version. Nothing should prevent use to do alpha and rc releases, and I don't see any "relation" to the API version, stability, compatibility, …
Doing a release, in alpha, or rc, is essentially stating, their might be some bugs, and code that will change, but *most likely* the featureset won't change. It's unrelated to APIs, it's just a matter of what we ship, when and with what level of confidence, isn't it ? 👼🏼

Reply:
I included information on software versioning here because I think it's easy to mix up software versioning with API versioning. This line was simply meant to be a description of the current state of the world, as I didn't see any "alpha", "beta", or "rc" versions when I looked through Tekton releases. This TEP is proposing to continue defining what is "breaking" or "backwards incompatible" based on the stability level of the API features, and to increment the software version accordingly based on semantic versioning. I am not sure I fully understand your question, but I hope this plus additional clarifying comments in the TEP answers it?

Reply:
It does and it doesn't at the same time 😛 . 

> I included information on software versioning here because I think it's easy to mix up software versioning with API versioning

In the k8s ecosystem, I am not sure that's true. I might be wrong here, but it seems very clear that the kubernetes object API version is orthogonal to the release version. API `v1` for Knative was published before Knative did a `v1` release (tagged v1), similar with internal k8s object.

> Instead of doing an "alpha" release, we would keep the API version at V1, add an alpha feature behind an alpha feature flag, and do a software minor version release.

This doesn't prevent us from doing alpha and rc releases though. We could release an v1.1.0-alpha release that provides  some new field behind an alpha feature-gate, and move to v1.1.0-rc and v1.1.0 final while still keeping this new field behind an alpha feature flag. This is the same upstream with Kubernetes tbh. The "alpha"/"rc" state of a given release is about the code, not about the API version and status, as we version APIs..

Reply:
Thanks for the feedback! I've made software versioning a non goal of this TEP.